Projectile Land, Projectile Tick triggers

neYaANK opened the pull request:
Working on #181
Made ProjectileLandTrigger that triggers when selected projectile hits the block, but i'm not really sure how to implement ProjectileTick as to do so i need to use at least more than one EventListener

Travja commented:
My suggestion for Projectile Tick would be to create a custom event for the ticking that you can use in the trigger. That way you can apply whatever logic outside of the trigger itself to determine if the projectile is moving and whatnot and just worry about trigger logic in the specific trigger itself.

neYaANK commented:
Okay so i've implemented ProjectileTickTrigger, was thinking about adding interval and delay for a trigger as setting but right now got no idea how to do so. Also TaskList's remove() method didn't cover all cases and caused a bug so i changed it too

Travja commented:
> Okay so i've implemented ProjectileTickTrigger, was thinking about adding interval and delay for a trigger as setting but right now got no idea how to do so. Also TaskList's remove() method didn't cover all cases and caused a bug so i changed it too

Great work! I'll get you some sample code on how to add those settings and things because I do feel like that would be important. A task running every tick can potentially be pretty expensive, especially if multiple projectiles are being fired and their child components have a bit going on.

neYaANK commented on src/main/java/com/sucy/skill/dynamic/trigger/ProjectileLandTrigger.java:
```diff
@@ -0,0 +1,70 @@
+     */
+    @Override
+    public LivingEntity getTarget(ProjectileHitEvent event, Settings settings) {
+        Location loc = event.getEntity().getLocation();
```
I did not make the ability for it to trigger upon entity hit as it is already handled by Took Physical Damage Trigger. Then i'll add it

Travja replied:
Took Physical Damage may not always apply. Some projectiles could be set up in such a way that they don't actually do damage.

neYaANK commented on src/main/java/com/sucy/skill/dynamic/trigger/ProjectileLandTrigger.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public boolean shouldTrigger(ProjectileHitEvent event, int level, Settings settings) {
+        if(!Objects.isNull(event.getHitEntity())) return false;
+            if (!(event.getEntity().getShooter() instanceof LivingEntity)) return false;
```
Right now it should work for all living entities, but not for dispensers. To be honest, i can't think of how we would use dispensers in a skill so i excluded them.

neYaANK replied:
Found out that this restriction in shouldTrigger was redundant as for Trigger to work caster must be a LivingEntity, so i removed it

Travja commented on src/main/java/com/sucy/skill/dynamic/trigger/ProjectileHitTrigger.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public boolean shouldTrigger(ProjectileHitEvent event, int level, Settings settings) {
+        List<String> projectiles = settings.getStringList("projectile");
+        String type = settings.getString("type","both");
```
It would probably be good to add `block` as an option here too. So we'd consider it hit on either block, entity, or both depending on the setting.

neYaANK replied:
Here actually if the `type` is anything but `both` or `entity`, it will be considered as a `block`. Should i add validation so `type` would accept only `entity`, `block` or `both` values?

neYaANK commented on src/main/java/com/sucy/skill/dynamic/trigger/ProjectileTickTrigger.java:
```diff
@@ -0,0 +1,62 @@
+    @Override
+    public LivingEntity getTarget(ProjectileTickEvent event, Settings settings) {
+        boolean targetCaster = settings.getBool("target",false);
+        return targetCaster? (LivingEntity) event.getEntity() : new TempEntity(new EntityTarget(event.getProjectile()));
```
`getProjectile()` returns `Projectile` which does not inherit `LivingEntity` interface so this hacky way is the only way that i found

Travja commented on src/main/java/com/sucy/skill/dynamic/trigger/ProjectileHitTrigger.java:
```diff
@@ -38,10 +39,15 @@ public Class<ProjectileHitEvent> getEvent() {
+        boolean hitBlock = !Objects.isNull(event.getHitBlock());
         boolean isCorrectProjectile = projectiles.isEmpty() || projectiles.contains("Any") || projectiles.stream().anyMatch(projectile->event.getEntityType().name().equalsIgnoreCase(projectile));
-        boolean isCorrectType = type.equalsIgnoreCase("both") || type.equalsIgnoreCase("entity") == hit;
+        boolean isCorrectType = type.equalsIgnoreCase("both") || (type.equalsIgnoreCase("entity") == hitEntity && type.equalsIgnoreCase("block") == hitBlock);
```
Shouldn't this be `|| type is entity || type is block` instead of `&&`?

neYaANK replied:
`type = "block"`
`hitEntity = false`
`type.equals("entity") == hitEnity` `false == false` and it will return true before reaching the `type is block` check

Travja replied:
I mean, I guess that works. The readability there is just very difficult.

neYaANK replied:
I just realized that this check is also redundant as i am validated the `type` value before so i removed it.

neYaANK commented:
The only thing left to be done is custom set delays and intervals for `ProjectileTickTrigger`, but I can't think of solutions other than having a `Map<Projectile,Integer>` where `Integer` is used for handling delays and intervals per each `Projectile` shot. Still, i don't find this solution appealing and would be really happy if you offer any other options

Travja commented:
Have you had a chance to actually test out these new triggers at all?

neYaANK commented:
Yeah, I test it every time before pushing to remote branch. Server I am using is Spigot 1.16.5

Travja commented:
Awesome. Things are looking good. I'll probably pull things down and make a couple formatting/readability adjustments before I hit the approve button. And I think I'd like to get that projectile tick timing setting set up as well before we merge this.

Do you know how to modify the editor at all to include these new options, or should I take care of that?

neYaANK commented:
I will take a look at editor after implementing interval setting, but not sure if i am capable of modifying editor.

neYaANK commented:
So, any feedback?

Travja commented on src/main/java/com/sucy/skill/listener/ProjectileListener.java:
```diff
@@ -0,0 +1,62 @@
+        if (!(shooter instanceof LivingEntity)) return;
+
+        flyingProjectiles.add(event.getEntity().getUniqueId());
+        MainThread.register(new ProjectileTickTask((LivingEntity) event.getEntity().getShooter(), event.getEntity()));
```
I'm a little weary of always scheduling the tick task here regardless of if it will be applicable at all. I kind of feel like if we were able to find a way to only run this task if there is a valid tick trigger on the firing player, that would be better. I'm not entirely sure the best implementation for this kind of thing, I'm just a little concerned about the potential performance hit that could be caused by tracking potentially dozens-hundreds of projectiles every tick like this.

Travja replied:
Yeah that's a good point. I'm not entirely sure the best solution at the moment, but it's worth giving a little more thought. At the end of the day, we may just go with exactly what you have and hope it doesn't cause too many issues with performance.

Travja reviewed (approved):
I'm going to go ahead and approve this. We'll see how that projectile tick measures up as people use it and we can make adjustments if it becomes an issue.
